### PR TITLE
[WIP] Initial cancellation proposal.

### PR DESCRIPTION
**Pull request by cheatfate**
This is initial Future[T] cancellation proposal and implementation.

* The most significant changes were made in `asyncfutures2.nim` and `asyncmacro2.nim`.

* `asyncloop.nim` has 2 examples on how cancellation can be handled for simple procedures (which are not transformed with `async` macro).

* `asyncfutures2.nim` also has most problematic places such as `all()` and `and()` which is marked with `# TODO` comments. I really think that this procedures needs some care or deprecation.

* Most of other places are just `if not future.finished(): future.finish/fail()` because when cancellation happens, you must not `complete()` or `fail()` Future[T].

* Also there present fix for recently discovered `EINTR + connect` problem.

### CODE REVIEW

**Comment by cheatfate** [comment]
The basic idea of using cancellation is graceful Future[T] shutdown at any moment of time. The simplest example is the execution of a specific operation/task with a
time limit or timeout. When timeout exceeded operation/task must be gracefully
closed (sockets/files or any other handles which are used must be closed).

Currently `chronos` using two types of asynchronous procedures. First type is
simple procedures which can wait for other async procedures only via closures and callbacks. Second type is transformed `{.async.}` procedures which are able to use `await` to wait for results of other async procedures.

First type of procedures can catch cancellation using newly introduced callback Future[T].cancelCallback. This callback will be called once specific Future[T] will be cancelled. Note that this callback could not be async procedure and could not wait other procedures using `await`.

```nim
proc sleepAsync*(duration: Duration): Future[void] =
  ## Suspends the execution of the current async procedure for the next
  ## ``duration`` time.
  var retFuture = newFuture[void]("chronos.sleepAsync(Duration)")
  let moment = Moment.fromNow(duration)

  proc completion(data: pointer) {.gcsafe.} =
    if not(retFuture.finished()):
      retFuture.complete()

  proc cancel(udata: pointer) {.gcsafe.} =
    ## Cancellation procedure is used to remove previously installed timer from
    ## the queue.
    if not(retFuture.finished()):
      removeTimer(moment, completion, cast[pointer](retFuture))

  retFuture.cancelCallback = cancel
  addTimer(moment, completion, cast[pointer](retFuture))
  return retFuture
```

Second type of procedures can catch cancellation using `try/except` block to catch `CancelledError` exception. Because exception handler is also placed inside of transformed `{.async.}` procedure it also can use `await`.

```nim
proc someAsyncProc*() {.async.} =
  try:
    var transp = await connect(someAddress)
    await transp.send("hello")
  except CancelledError:
    ## You can place here asynchronous cancellation handler, which is able to
    ## use `await`.
    if not(transp.closed):
      await transp.closeWait()
```

File: chronos/asyncmacro2.nim [hunk 1]
```diff
@@ -81,11 +83,19 @@ template useVar(result: var NimNode, futureVarNode: NimNode, valueReceiver,
 
 template createVar(result: var NimNode, futSymName: string,
                    asyncProc: NimNode,
-                   valueReceiver, rootReceiver: untyped,
+                   valueReceiver, rootReceiver, retFutSym: untyped,
                    fromNode: NimNode) =
   result = newNimNode(nnkStmtList, fromNode)
   var futSym = genSym(nskVar, "future")
   result.add newVarStmt(futSym, asyncProc) # -> var future<x> = y
+  # retFuture.child = future<x>
```
Comment (yglukhov) [inline chronos/asyncmacro2.nim]: I think all of this template can greatly benefit from using `quote do` notation

Comment (cheatfate) [reply chronos/asyncmacro2.nim]: I even do not understand why this templates not compile-time procedures?

File: chronos/asyncfutures2.nim [hunk 2]
```diff
@@ -480,6 +538,11 @@ proc all*[T](futs: varargs[Future[T]]): auto =
   ##
   ## Note, that if one of the futures in ``futs`` will fail, result of ``all()``
   ## will also be failed with error from failed future.
+  ##
+  ## TODO: This procedure has bug on handling cancelled futures from ``futs``.
```
[the remark this answers is not in the document]
Comment (cheatfate) [reply chronos/asyncfutures2.nim]: This `all()` is going to be deprecated because it is logically broken. From my point of view all the aggregation procedures which are returning actual values are logically broken. Because it hard to understand what you need to return as aggregated data value, when one of the futures got failed or cancelled.

File: chronos/asyncfutures2.nim [hunk 3]
```diff
@@ -397,75 +448,82 @@ proc mget*[T](future: FutureVar[T]): var T =
...
-
 proc asyncCheck*[T](future: Future[T]) =
   ## Sets a callback on ``future`` which raises an exception if the future
   ## finished with an error.
   ##
   ## This should be used instead of ``discard`` to discard void futures.
   doAssert(not isNil(future), "Future is nil")
   proc cb(data: pointer) =
-    if future.failed:
+    if future.failed() or future.cancelled():
```
[the remark this answers is not in the document]
Comment (cheatfate) [reply chronos/asyncfutures2.nim]: Absolutely agree with deprecation of `asyncCheck`, but could not agree with treating `cancellation` not as an error. Cancellation is not a normal code flow, its exceptional way of doing things to avoid leaks and other problems.

**Comment by disruptek** [comment]
I would argue that if you've written any code to handle cancellation, then it is by definition an acceptable code path and clearly not exceptional.  Exceptions should be used for behavior we cannot predict, not behavior we might usefully write code paths against.  To state that differently, exceptions should not be used for control flow.

**Comment by disruptek** [comment]
Also, unless I'm missing something, a cancelled future should, if possible, not be allowed to complete, because the user always has the option to produce identical behavior using existing means -- simply disregard the result of a future.  On the other hand, cancellations yield new suspension behavior that we cannot easily implement otherwise.

File: chronos/asyncfutures2.nim (hunk 3 above)
[the remark this answers is not in the document]
Comment (yglukhov) [reply chronos/asyncfutures2.nim]: Well in my world cancellation is a perfectly normal codeflow. Also it's a neat natural way to break spawned loops. But thinking further about this particular case, I conclude that it's nearly impossible to hit `cancelled` condition in `asyncCheck`, because usually `asyncCheck` implies that you don't have the reference to the future, and thus will not be able to cancel it, so I no longer care :)

**Comment by yglukhov** [comment]
> a cancelled future should, if possible, not be allowed to complete

I'd say a cancelled future should be impossible to read (raises CancelError), but IMO it should be fine to allow to complete it. Because:
```nim
proc doSmth(): Future[int] =
  let fut = newFuture[int]()
  scheduleSomeOperationInRunLoop() do(res: int):
    fut.complete(res) # Previously perfectly fine, but now I have to always check for cancelled state or else...
  fut
```
Now if I forget to add cancelled-state-check (and I will) I will not notice any bug until my future is cancelled. If my future is cancelled on e.g. timeout, this will lead to basically a heisenbug, which by all laws will not be reproducible on my machine, only my users will be affected :).

So to summarize, `complete` should be callable as it was, but if the future is cancelled, `complete` should be a noop. Also `complete` should never be called more than once.

File: chronos/asyncfutures2.nim (hunk 2 above)
[the remark this answers is not in the document]
Comment (yglukhov) [reply chronos/asyncfutures2.nim]: `all` is synonymous to `and` to me. So ideally it should return a corresponding tuple (i.e. `proc all(f1: Future[T1], f2: Future[T2], ...): (T1, T2, ...)`) if `all` futures are completed, and fail if at least one of the futures has failed. That said, I see it more as a metaprogramming exercise rather than something really useful :)

**Comment by cheatfate** [comment]
@kdeme @yglukhov Future cancellation can be happen at any time, so when you receiving your callback you actually do not know about state of your `Future[T]`. So without this check you can try to move Future[T] state from `CANCELLED` -> `COMPLETED` or `FAILED`.

Also because callbacks do not have strong relation with Future[T], e.g. callback's `udata` argument is not always Future[T] address, we can't deprecate/remove callbacks from poll()'s queue.

So callback will be dispatched even if related Future[T] was cancelled, with this check we are trying to avoid completion code to be executed.

**Comment by cheatfate** [comment]
@kdeme cancellation happens from above, e.g. parent cancelling child. So for example `someAsyncProc()` can be called in the way like this:
```nim
  var timeoutFut = sleepAsync(10.seconds)
  var someFut = someAsyncProc()
  await someFut or timeoutFut
  if timeoutFut.finished():
    someFut.cancel()
  else:
    timeoutFut.cancel()
```
In such way branch `except CancelledError` will be executed.

**Comment by yglukhov** [comment]
> Future cancellation can be happen at any time, so when you receiving your callback you actually do not know about state of your Future[T]. So without this check you can try to move Future[T] state from CANCELLED -> COMPLETED or FAILED.

My point is exactly that. `proc complete` should do the state check internally and do nothing if state is `CANCELLED`.

**Comment by cheatfate** [comment]
@yglukhov but before procedure call to `Future[T].complete()` callback can perform reading on socket/file/pipe or writing to socket/file/pipe and only after that it will call to `retFuture.complete()`. The reason of `finished` checks is to prevent the execution of this code.

File: chronos/asyncfutures2.nim [hunk 4]
```diff
@@ -365,30 +415,31 @@ proc injectStacktrace[T](future: Future[T]) =
 
 proc read*[T](future: Future[T] | FutureVar[T]): T =
   ## Retrieves the value of ``future``. Future must be finished otherwise
-  ## this function will fail with a ``ValueError`` exception.
+  ## this function will fail with a ``FutureError`` exception.
```
[the remark this answers is not in the document]
Comment (cheatfate) [reply chronos/asyncfutures2.nim]: Reverted everything back to `ValueError`.

**Comment by cheatfate** [comment]
@yglukhov @kdeme @arnetheduck my proposal is to deprecated `and`, `or`, `all`, `oneIndex`, `oneValue` and introduce only this 2 procedures:

```nim
proc all*[T](futs: varargs[Future[T]]): Future[void] =
  ## Returns a future which will complete only when all futures in ``futs``
  ## will be completed, failed or canceled.
  ##
  ## If the argument is empty, the returned future COMPLETES immediately.
  var retFuture = newFuture[void]("chronos.all()")
  let totalFutures = len(futs)
  var completedFutures = 0

  # Because we can't capture varargs[T] in closures we need to create copy.
  var nfuts = @futs

  proc cb(udata: pointer) {.gcsafe.} =
    if not(retFuture.finished()):
      inc(completedFutures)
      if completedFutures == totalFutures:
        retFuture.complete()

  proc cancel(udata: pointer) {.gcsafe.} =
    if not(retFuture.finished()):
      for i in 0..<len(nfuts):
        if not(nfuts[i].finished()):
          nfuts[i].removeCallback(cb)
          nfuts[i].cancel()

  for fut in nfuts:
    fut.addCallback(cb)

  retFuture.cancelCallback = cancel
  if len(nfuts) == 0:
    retFuture.complete()

  return retFuture

proc one*[T](futs: varargs[Future[T]]): Future[int] =
  ## Returns a future which will complete and return index inside of ``futs``
  ## when one of the futures in ``futs`` will be completed, failed or canceled.
  ##
  ## If the argument is empty, the returned future FAILS immediately.
  ##
  ## On success returned Future will hold index in ``futs`` array.
  var retFuture = newFuture[int]("chronos.one()")

  # Because we can't capture varargs[T] in closures we need to create copy.
  var nfuts = @futs

  proc cb(udata: pointer) {.gcsafe.} =
    if not(retFuture.finished()):
      var res = -1
      var rfut = cast[FutureBase](udata)
      for i in 0..<len(nfuts):
        if cast[FutureBase](nfuts[i]) != rfut:
          nfuts[i].removeCallback(cb)
        else:
          res = i
      retFuture.complete(res)

  proc cancel(udata: pointer) {.gcsafe.} =
    if not(retFuture.finished()):
      for i in 0..<len(nfuts):
        if not(nfuts[i].finished()):
          nfuts[i].removeCallback(cb)
          nfuts[i].cancel()

  for fut in nfuts:
    fut.addCallback(cb)

  if len(nfuts) == 0:
    retFuture.fail(newException(ValueError, "Empty Future[T] list"))

  retFuture.cancelCallback = cancel
  return retFuture
```

**Comment by arnetheduck** [comment]
Both of these functions require multiple ownership of the futures - take `one` for example, it will return the index into the list of futures passed into the function - practically, this means that both the calling code and the closures must keep references around - this will eventually turn into a bottleneck that can't be worked around. 

Can you think of any designs which avoid this issue? This would make the API more composable, so that the futures "flow" through the layer but have fewer owners (with the goal of having just a single owner, which is typically done by separating writing to the future from reading from it - there should ever only be one "Writer")

**Comment by cheatfate** [comment]
@arnetheduck result of `one` can be `Future[Future[T]]` if you do not like index. Are you agree with `all()`.

**Comment by kdeme** [comment]
If I understand it right, this `all` version will not return failed if any of its futures fails? So an exception will not be propagated. E.g. this code uses this behaviour: https://github.com/status-im/nim-eth/blob/master/eth/p2p/rlpx.nim#L904

Next to that, the `all` version that returns values is used in kademlia.nim on 3 locations. Perhaps this proves that there is some case for having this?

If so I'd say the suggestion of @yglukhov is most intuitive to me (return all values in a seq or fail when 1 fails, but still wait till all complete/fail/cancel)

I haven't given this much thought yet, so it might be crazy/inefficient/unworkable, but would it be possible to return a future with a `seq[Future[T]]`? And then some helper functions e.g. `fold` or `readAll` that reads all the futures and raises on failed? But perhaps this is not really needed.

**Comment by zah** [comment]
When you say:

```nim
1) var results = await all(foo(), bar())
2) #do something with the results
```

I think the semantics should be the following:

The code execution should reach line 2 only if the async operations `foo()` and `bar()` were able to successfully produce their results. This mimics the semantics of non-async code and thus it can express any algorithm that is expressible with non-async code. Failed and cancelled futures are both failures to reach these post-conditions and can be handled with exceptions.

The new support for cancellation gives us a new trick we can add to the API. If any of the futures fails or is cancelled, the framework can immediately try to cancel all other outstanding futures. Any installed callbacks can be notified in this case.

Please note that the alternative treatment suggested by @cheatfate is already available if you use `yield` instead of `await`. Then you'll get a sequence of future objects that may be in failed or cancelled state. More info here:

https://nim-lang.org/docs/asyncdispatch.html#asynchronous-procedures-handling-exceptions

The above semantics of `all` mimic closely what we can have in the future with [enforced awaits](https://github.com/status-im/nim-chronos/issues/2) (the example above is equivalent to `var results = await(foo(), bar())` and they get us on the track towards having a single owner as suggested by @arnetheduck.

The proposed alternative `select` API is a nice way to side-step the ownership problems when using APIs such as `one`. Besides the ownership issues, `one` also has a type-safety issue. If you pass a heterogenous list of futures to `all`, the returned type from `all` cannot be strongly typed, so some manual upcasting is required.

**Comment by cheatfate** [comment]
So with latest changes i decided not to perform automatic cancellation of Futures in `all()`. Because all this Futures was created in **invoker**, so if `all()` will be cancelled its **invoker** problem to cancel all the created Futures.

**Comment by cheatfate** [comment]
@kdeme please explain me what the difference between Future[T], Option[T] and Result[T] in terms of error handling? There no difference, maybe API different, but all three provides equal functionality.

**Comment by cheatfate** [comment]
@kdeme and @yglukhov proposing too keep in code undefined behavior.
Even if you keep waiting for all the Futures[T] to be completed/cancelled/failed you how you can choose what exception will be raised? First in list? First which appear? Maybe Futures you passed to `all()` are not equal? With what error code you should fail your result Future[T]?
My vision, is that `chronos` must not decide on this, this is your Future[T] list and aggregation procedures must only wait until something happens. And its your own code must check if some of the Futures in your list failed/cancelled/completed and log errors or do something useful.

**Comment by cheatfate** [comment]
@kdeme and this is not hiding error, it is transferring error condition to your side, so you will make choice how to react on success and failure.

**Comment by zah** [comment]
I think reporting the first failure as an exception is a reasonable compromise. The other futures get cancelled immediately and if the code later tries to fail or complete them, this is a no-op.

Again, you can still use `yield` if that's what you want.

**Comment by cheatfate** [comment]
@zah i have changed this behavior, so all the Future[T] will not be cancelled anymore.

**Comment by cheatfate** [comment]
The only procedures which are now cancelling automatically its `withTimeout()` and `wait`.

**Comment by cheatfate** [comment]
@zah i dont think `chronos` must decide what error should be returned from the list of errors, and in most of the cases this error will be absolutely random. Its fine if list passed to `all()` consist of homogeneous Future[T], but what if list is heterogeneous?

**Comment by zah** [comment]
I don't see this introduction of a slight non-determinism as a show-stopper. As an example, consider that in C/C++ the order of evaluation of function arguments is implementation-defined. This means that if you have code like `foo(bar(), baz())` where both `bar()` and `baz()` can fail, you don't know which exception will be raised. This haven't proven much of a problem for writing synchronous algorithms, so we can probably live with it in the async world as well.

Errors can be returned though either error codes or exceptions. If all of the functions return an error code successfully, this doesn't violate the await statement post-conditions and execution will continue normally. If even one of the calls raises an exception, this exception is the result of the whole `await` statement.

**Comment by zah** [comment]
I think the main distinction between our views is that @cheatfate wants an API that is able to communicate as much information as possible regarding what happened, so you can express a great variety of async algorithms. My view is that we should have a long-term goal of restricting the possible async algorithms further in order to gain the optimisations and new functionality described in the enforced awaits RFC.

**Comment by kdeme** [comment]
> @kdeme please explain me what the difference between Future[T], Option[T] and Result[T] in terms of error handling? There no difference, maybe API different, but all three provides equal functionality.

What I meant here is that perhaps a returned sequence of results could be a good solution? But as stated before, a sequence of futures could also be done, yes, functionality remains similar.

> @kdeme and @yglukhov proposing too keep in code undefined behavior.

I very much agree that `all` can be interpreted ambiguously, hence the problem as you describe it. 
In that sense, the now `allFutures` is less ambiguous, but with the cost of simply removing functionality. Perhaps that is better than some unclear implementation, but perhaps a clear description in the documentation for `all` would be fine too.

So, I'm not necessarily against this change, but I'm just curious if we will not just end up rewriting another `all` to fit our needs. And perhaps that is OK, as we will see in a more pragmatic way what we really need in the code. And perhaps that can lead to a more "useful" implementation of `all`.

So I gave this a quick try, and as I see it, https://github.com/status-im/nim-eth/blob/master/eth/p2p/rlpx.nim#L904, with same functionality could become:
```nim
await allFutures(subProtocolsHandshakes)
for f in subProtocolsHandshakes:
  if f.failed:
    f.read()
```
(This is perhaps not the best example, as if we need to do this I would probably do a proper calling of the disconnect handlers here also)

https://github.com/status-im/nim-eth/blob/master/eth/p2p/kademlia.nim#L385 could become something like:
```nim
let futs = candidates.mapIt(k.bond(it))
await allFutures(futs)
for i in 0 ..< futs.len:
  let bonded = futs[i].read()
  if not bonded: candidates[i] = nil
```
Is this how you have it envisioned @cheatfate ? Or am I using it wrongly?
I see two drawback with this:
1. The previously mentioned ownership, however practically, they are not a big issue in these 2 cases.
2. The fact that you can perfectly replace `await` with `yield`, as `allFutures` will not raise. This basically turns down to what @zah said that you can already accomplish this with `yield`.

So yeah, all in all ( :-)) perhaps the `all` which fails is not so bad yet? 

The point that @zah makes regarding enforced awaits does make sense to me from a quick view, but I still have to check the RFC in better detail.

**Comment by arnetheduck** [comment]
> lease explain me what the difference between Future[T], Option[T] and Result[T]

the difference would be that future leaves open the possibility that it still lacks a value after `all` (even if it "shouldn't"), from an api perspective, whereas option/result clearly and unambiously describe that the value or error exists _now_, meaning there's one less opportunity to misuse the api

**Comment by cheatfate** [comment]
@zah when we implement your RFC we don't need this `all()` and `one()` function at all, but you still blocking me from merging this PR and moving forward.

**Comment by arnetheduck** [comment]
for now I'd leave the RFC aside unless there are clear and easy partial wins in it that we can apply already - I wouldn't consider it a blocker for this API, which we can grow organically from here. if we run into issues with it, it's cheap to change - we don't currently guarantee backwards compat in any shape and have deps locked down. refactoring is easy.

**Comment by cheatfate** [comment]
@arnetheduck @kdeme you trying to use **actual value** as result from calling function, but in asynchronous world you are getting only **promise/future** and not **actual value**. And when you waiting all the **promises/futures** you waiting when they become completed, so they actually have **value** or **error**. So you trying to work with **result values** while i trying to work with **promises/futures**. And after all the **promises/futures** got completed, you can use whatever API you want to gather **result values** from it.
`Option[T]` and `Result[T]` are made to work in sync world, when **result value** or **error value** available right after procedure/function call. And so this primitives missing state when **result value** is still missing. In async world it happens when you trying not to use `await someProcedure()`, but using `var res = someProcedure`. And even in such way API is not misused, because your **res** is not a value, its Future[T] which holds your value, and to read it you need to call **res.read()**, which will produce exception/error if there no value.

**Comment by cheatfate** [comment]
@kdeme your code examples are still generating exceptions, because `read()` call is generating exception if Future[T] is either **NOT READY**, **CANCELLED** or **FAILED**.

**Comment by arnetheduck** [comment]
>  work in sync world

indeed - one could imagine that `await all(futures)` could return a seq of `result` so that the errors and successes can be individually assessed, instead of having a single raise for the whole thing. 

an outcome of this would be that `await all(futures)` would "never" raise which would avoid some annoying double error handling.

**Comment by kdeme** [comment]
> @kdeme your code examples are still generating exceptions, because `read()` call is generating exception if Future[T] is either **NOT READY**, **CANCELLED** or **FAILED**.
> 
> So if you like to get an exception, you just need to call `read()`.

Yes, I know. I did that on purpose as my point was trying out your updated `all` while keeping same behaviour.

**Comment by arnetheduck** [comment]
seq or tuple, depending on the homogenity of the arguments passed to `all`, that is.. imagine:
`let (a, b, c) = await all(afut, bfut, cfut)` for example..

it would be a `result[T, SomeException]`, and it would probably make sense that the dereference operator of `result` raises the given exception if it's set, instead of raising some other generic reception

**Comment by cheatfate** [comment]
@kdeme for example you are trying to wait for **N** HTTP request and you need to update some table with results of this requests, is it possible to finish with `all()` you trying to have?
Or it can happens that one error could mask all other errors:
1. Failed because hostname could not be resolved (e.g. DNS error) -> RETURNED.
2 - 10. Failed because we lost internet connection while receiving responses. -> LOST.
And if you know that connection is lost you don't need to continue because it fatal, but you got different error, and so will try to continue in fatal behavior.

**Comment by cheatfate** [comment]
@arnetheduck sorry but why we running in circle really. `result[T, SomeException]` is still absolutely equal thing. Because to get **result value** from this **result** you need to call `result.get()/read()/value()`, in Futures code you need to call `future.read()` to get **result value**. So what the difference?
```nim
let (a, b, c) = await all(afut, bfut, cfut)
let avalue = a.value()
let bvalue = b.value()
let cvalue = c.value()
```

```nim
let (a, b, c) = await all(afut, bfut, cfut)
let avalue = a.read()
let bvalue = b.read()
let cvalue = c.read()
```

**Comment by arnetheduck** [comment]
The difference that the API of result guarantees that either a value or an error exists - there can be no doubt. if you receive a `future` from an API, it admits the possibility that the value still doesn't exist (if you look at the future in isolatio, forgetting that it passed through "all" - as may happen when you pass through many layers). from a clarity perspective of what `all` does, switching from future to result in the final type makes it unambiguous...

using type this way would for example prevent that you await a future twice, as well as other silly mistakes developers do - async is hard to reason about, doubly so when it's hidden behind the macro magic.

**Comment by cheatfate** [comment]
@arnetheduck its why i wish to avoid API which can raise disputes like this. Because today you all against my API, while tomorrow picture can be absolutely different. While my approach has more logic, your approach can be viewed from different angles and depend on angle it can be either right or wrong.

**Comment by arnetheduck** [comment]
one possible way to achieve this is to have something like:
```
proc all[T](openarray[future[T]): future[seq[T, Exception]] =
  ...
```

notice how it's no longer necessary to document what `all` does - it's clear and obvious that `all` returns a future, and when that future completes, the values *will* be available.

**Comment by cheatfate** [comment]
Chronos API must not operate on **result values**, after all this discussions this my core opinion and you guys all strengthened it in my mind. If you wish to get **result values** call `read()`.

**Comment by cheatfate** [comment]
@arnetheduck the only possible way to achieve is to start work on @zah RFC. Which can be used to safely organize **result values**. Right now to stop all this disputes i wish to leave this `all/or/and` as it was before, and allow to use it and start introducing exceptions and undefined behavior to just merge this PR and go forward. The only thing which i have tried to made is to remove `undefined behavior` code, but for some reason you are protecting this parts.

**Comment by arnetheduck** [comment]
that's fine too, my suggestions are not blockers - we can refactor this at any time or even keep this code outside of chronos (at a tiny performance loss) .

since you asked for feedback, I'm just showing you what an API that does not lose any information could look like:
* it keeps track of all values or exceptions individually
* it uses type to keep track of which results have been completed, so as to avoid deadlocks, double future completions etc

once you have this, it's easy to build other forms of aggregation on top.

**Comment by cheatfate** [comment]
@arnetheduck there no difference with such line and it much more natural to async and also introduce absolutely equal behavior:

```proc all[T](futs: varargs[Future[T]]): Future[seq[Future[T]]```

**Comment by arnetheduck** [comment]
> absolutely equal behavior:

not quite - for `seq[Future[T]]` the following is valid:
```
let s = await all(...)
await s[0]
```

you'll agree that this is a nonsensical operation? with `result`, that gives a compile error.

**Comment by arnetheduck** [comment]
or even:
```
let s = await all(...)
s[0].complete(..)
```

I've seen bugs like this in the wild :)

**Comment by cheatfate** [comment]
@arnetheduck sorry but my API did not loose any information, `Future[T]` you passed to `all()` are still has all the error codes or values after `all()` call.

Also deadlocks and double completion never happens in linear code without parallel running tasks. You can have problems with dead locks and race conditions in multithreaded application if you not using proper synchronization primitives. I have made such primitives year ago, but because people do not have experience with using it they started to use `variables` as flags and they thought they are safe...

**Comment by cheatfate** [comment]
@arnetheduck try to use it and you will get runtime errors in both cases about double completion.

**Comment by arnetheduck** [comment]
indeed - looks like we're in agreement: with `Future` you  get runtime errors. `Result` allows you to detect this error at compile time instead - this is the information "loss", from a type perspective.

**Comment by cheatfate** [comment]
If you want it will be easy to add such check for Future[T], so you will not have ability to receive value from Future[T] twice.

**Comment by cheatfate** [comment]
@arnetheduck of course we get **runtime** errors because we have not 2 states but at least 4 states, so Future[T] is more general case for Result[T].

**Comment by arnetheduck** [comment]
well, it is the job of `all` to reduce to 2 states (assuming there are no bugs in the implementation of `all`!) - as an API designer, you can either choose to encode this reduction in the type, or discard it and keep using the future.

**Comment by cheatfate** [comment]
@arnetheduck also you manipulating
```
import result

type R = Result[int, string]

# Basic usage, producer
proc works(): R =
  R.ok(42)

when isMainModule:
  let a = works()
  echo a.value
  echo a.value
```
This works and do not have any problems at compile time as well as at runtime.

**Comment by cheatfate** [comment]
@arnetheduck the job of `allFutures()` is to jump to time when all the futures in list will be **finished**. And term **finished** can be explained as (**failed** or **completed** or **cancelled**).

**Comment by arnetheduck** [comment]
> This works and do not have any problems at compile time as well as at runtime.

yes, you're right, neither Result nor Future restrict the number of times you can read the value - I don't think this has been disputed anywhere.

**Comment by cheatfate** [comment]
Also i named it `

> 
> 
> > This works and do not have any problems at compile time as well as at runtime.
> 
> yes, you're right, neither Result nor Future restrict the number of times you can read the value - I don't think this has been disputed anywhere.

https://github.com/status-im/nim-chronos/pull/41#issuecomment-508542376

**Comment by arnetheduck** [comment]
> #41 (comment)

the point of that comment is that you can `await` twice - you can't `await` a `result` - there's nothing to wait for any more.

**Comment by cheatfate** [comment]
> 
> 
> > #41 (comment)
> 
> the point of that comment is that you can `await` twice - you can't `await` a `result` - there's nothing to wait for any more.

```nim
import chronos

proc testFuture(): Future[int] {.async.} =
  await sleepAsync(100.milliseconds)
  result = 10

proc testDoubleAwait() {.async.} =
  var fut = testFuture()
  var r1 = await fut
  echo r1
  var r2 = await fut
  echo r2

when isMainModule:
  waitFor testDoubleAwait()
```

and 

```nim
import result

type R = Result[int, string]

# Basic usage, producer
proc works(): R =
  R.ok(42)

when isMainModule:
  let a = works()
  echo a.value
  echo a.value
```

**Comment by arnetheduck** [comment]
yes, so we agree both work. but I would go ahead and say that repeating `await fut` is silly. the second time, it would be guaranteed safe to `var r2 = fut.read()`. using `await` there is semantically confusing - why are we waiting again?

**Comment by cheatfate** [comment]
> 
> 
> yes, so we agree both work. but I would go ahead and say that repeating `await fut` is silly. the second time, it would be guaranteed safe to `var r2 = fut.read()`. using `await` there is semantically confusing - why are we waiting again?

This analysis can be done only inside of `await` macro in case of Nim. In case of other languages it is part of the compiler.

**Comment by kdeme** [comment]
To be clear, I'm fine with either keeping or removing the `all`. I realise it is not a perfect solution, but neither is having *only* `allFutures` in my opinion. But, we can work around it, so no big problem.

The points I was trying to make are:
- I recognize the ambiguity with the current `all`
- I think the `allFutures`, while being more clear, will not be enough for all scenarios. For sure there are cases where this is sufficient. Just as there are where now `all` can be used. And just as there probably are case where both are a bad fit.
- `allFutures` can already be achieved with yield on `all`. 
- And then, perhaps I was not clear enough on this, but my preference leans towards changing the current `all` that returns values into the direction of a `proc all[T](futs: varargs[Future[T]]): Future[seq[Future[T]]` or `proc all[T](openarray[future[T]): future[seq[result[T, Exception]]]`. As there can be no ambiguity about this.
The first was what I was initially suggesting, the second I realised afterwards could be a nice fit (hence the edit). But I guess this boils down to whether you want your code to continue with result or with exceptions (although both can work nicely when using result here as explained by @arnetheduck).
And I'm not saying that we have to block this PR for this, it can be done in the future if there deems a need for it.
Regarding the `Future` versus `result` there, I agree with @arnetheduck that the latter creates an API that is less easy to "abuse". Additionally, it actually might be a good/convenient location to move from exceptions to results.
But sure, no real need for this now, and you can also start with returning the futures and write extra sugar to convert them to results if you like that, but it will introduce another loop. Also, it doesn't necessarily have to be one or the other.

Anyway, that's my view for now. Disclaimer: it can be altered by future experiences & opinions :)

**Comment by arnetheduck** [comment]
If we want to differentiate "failed" from "cancelled", `result` would likely not be enough - an even more "pure" solution then would be a third type `CompletedFuture` or whatever that can make the distinction without loss.. that said, with exceptions, you can usually let the exception types leak between layers and get away with it because practically you're will not be catching and translating them at each level anyway (you probably won't be catching them at all...), so a `CancelledError` can also be used..

**Comment by cheatfate** [comment]
Sorry guys but this attempts to move from `Future[T]` to `Result[T]` translation from `Future[T]` to `Result[T]` is just redundant work.

`Result[T]` supports only 2 states, usually it is used to get result from function/procedure call. This 2 states are: `Result[T]` is **FAILED** or `Result[T]` is **COMPLETED**. Every attempt to read value from `Result[T]` when `Result[T]` is in **FAILED** state will raise an exception. Also you have some utilities to check state of `Result[T]`.

`Future[T]` supports 4 states, usually it is used to get result from function/procedure call. This 4 states are: Future[T] is **FAILED**, Future[T] is **COMPLETED**, Future[T] is **PENDING**, Future[T] is **CANCELLED**. Every attempt to read value from `Future[T]` when `Future[T]` is in **PENDING**, **CANCELLED** or **FAILED** state will raise an exception. Also you have some utilities to check state of `Future[T]`.

So `Future[T]` is more general version of `Result[T]` which supports result values which can be give in FUTURE, while `Result[T]` supports only result values which can be obtained immediately.

**Comment by cheatfate** [comment]
@kdeme Please, note that when we merge this PR, you can't use `yield` anymore. You of course can use it its not prohibited, but you will break **CANCELLATION** support for this procedures.

```nim
var x = await someProc()
```

was transformed to

```nim
var fut = someProc()
yield fut
var x = fut.read()
```

But if we accept this PR this call will looks like:

```nim
var fut = someProc()
retFuture.child = fut # <--- this was added.
yield fut
var x = fut.read()
```

In such way we are keeping linked list of Future[T] and it is used to cancel all the list. Because {.async.} transformed procedures do not have access to `retFuture`, you can't simulate proper behavior with `yield` anymore.

**Comment by zah** [comment]
No problem for me for this to be merged as well. 

My main point was that the current treatment of `all` is closer to the patterns we want in the future, so deprecating it is a bit too harsh. The other point was that the equivalent of `allFutures` is already available when you use `yield` instead of `await` (@cheatfate, I haven't seen a reply from you regarding this point). Otherwise, I would acknowledge that there are some finer points regarding auto-cancellation that are up for disputing.

Jacek's Result-based API can be called `try_await` or something similar in the future and it can provide an alternative for situations where you don't want any exceptions to be raised. Eugene seems to want to avoid spending time on any such type-safe macros at the moment, but these would be relatively easy to create actually. I'm volunteering to implement them in the near future.

**Comment by cheatfate** [comment]
@zah i just think to add `awaitne` (await No Exceptions) which will replace `yield` primitive.

**Comment by arnetheduck** [comment]
just for the record, the goal of using result in my suggestion wasn't about avoiding exceptions - it was just to not lose information when going from a single future being completed to multiple.. result happens to be a type that can hold sufficient information (if you consider cancelled a failure, which pretty much it is - no value will ever come)
